aip/auth/4122.md

```diff
@@ -0,0 +1,349 @@
+```
+
+The request JSON body **must** be of the form:
+
+```json
+{
+  "delegates": [
+    // optional list of delegates
+  ],
+  "scope": [
+    // required list of scopes
+  ],
+  "lifetime": "lifetime"
+}
+```
+
+The following is an example of the payload for a direct request for an access
+ token with one scope and a lifetime of 300 seconds:
+
+```json
+{
+  "scope": [
+    "https://www.googleapis.com/auth/cloud-platform"
+  ],
+  "lifetime": "300s" 
+}
+```
+
+The following is an example of the payload for a delegated request for an access
+token with one scope and a lifetime of 300 seconds:
+
+```json
+{
+  "delegates": [
+    "sa-delegate1@project-id.iam.gserviceaccount.com",
+    "sa-delegate2@project-id.iam.gserviceaccount.com"
+  ],
+  "scope": [
+    "https://www.googleapis.com/auth/cloud-platform"
+  ],
+  "lifetime": "300s"
+}
+```
+
+### Generating OpenID Connect ID tokens
+
+#### Parameters
+
+The auth library **must** implement the language-specific abstractions required
+to allow a caller to specify the following parameters for a request to generate
+an Open ID Connect ID token on behalf of a service account.
+
+**Source Credentials** - **Required**
+
+The library specific representation of the service/user account requesting
+ the impersonation.
+ 
+**Target Principal** - **Required**
+
+The email address of the service account to impersonate, e.g. 
+`sa-name@project.iam.gserviceaccount.com`.
+
+**Audience** - **Required**
+
+The value of the `aud` field in the ID token.
+
+**Delegates** - **Optional**
+
+An ordered list of service account emails in the delegation chain, if used.
+
+**Include Email** - **Optional**
+
+Whether or not to include the target principal email in the generated ID token.
+
+#### Expected Behavior
+
+Auth libraries **must** send a POST request to the 
+[projects.serviceAccounts.generateIdToken][3] method, using an access
+token generated for the Source Credential to authorize the request:
+
+```
+POST https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/<TargetPrincipal>:generateIdToken
+```
+
+The request JSON body **must** be of the form:
+
+```json
+{
+  "delegates": [
+    // optional list of delegates
+  ],
+  "audience": "audience",
+  "includeEmail": "true|false"
+}
+```
+
+The following is an example of the payload for a direct request for an id token
+with email included:
+
+```json
+{
+  "audience": [
+    "example-sa@example-project.iam.gserviceaccount.com"
+  ],
+  "includeEmail": true 
+}
+```
+
+The following is an example of the payload for a delegated request for an id 
+token with email excluded:
+
+```json
+{
+  "delegates": [
+    "sa-delegate1@project-id.iam.gserviceaccount.com",
+    "sa-delegate2@project-id.iam.gserviceaccount.com"
+  ],
+  "audience": "example-sa@example-project.iam.gserviceaccount.com"
+}
+```
+
+### Creating a self-signed JSON Web Token (JWT)
```

Comment:
Are there other use-cases besides user impersonation here? If not should we tailor implementations more to that use case. We already generate the iamcredentials client libraries if users want to just call these endpoints.